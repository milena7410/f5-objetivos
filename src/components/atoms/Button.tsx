import React from "react";
import { Pressable, PressableProps, Text, View } from "react-native";

type ButtonProps = {
  title: string;
  variant?: "primary" | "secondary" | "danger";
  disabled?: boolean;
  textClassname?: string;
  buttonClassname?: string;
} & PressableProps;

type Variant = {
  disabled: { primary: string; secondary: string; danger: string };
  enabled: { primary: string; secondary: string; danger: string };
};
const Button: React.FC<ButtonProps> = ({
  variant = "primary",
  title,
  disabled,
  className = "",
  textClassname = "",
  buttonClassname = "",
  ...rest
}) => {
  const variantButtonStyle: Variant = {
    enabled: {
      primary: " bg-black dark:bg-white",
      secondary: "border border-black bg-white dark:bg-black dark:border-white",
      danger: "bg-danger",
    },
    disabled: {
      primary: "bg-gray-500",
      secondary: "bg-gray-300",
      danger: "bg-red-300",
    },
  };

  const textStyle: Variant = {
    enabled: {
      primary: "text-white dark:text-black",
      secondary: "text-black dark:text-white",
      danger: "text-txt-danger",
    },
    disabled: {
      primary: "text-white",
      secondary: "text-black ",
      danger: "text-txt-danger",
    },
  };

  const disabledProp = disabled ? "disabled" : "enabled";
  return (
    <Pressable
      className={`${buttonClassname} rounded-full self-center h-14 w-40 "`}
      disabled={disabled}
      {...rest}
    >
      <View
        className={`${className}  ${variantButtonStyle[disabledProp][variant]} items-center justify-center rounded-full flex-1 self-stretch`}
      >
        <Text
          className={`${textClassname} text-sm font-bold ${textStyle[disabledProp][variant]}`}
        >
          {title}
        </Text>
      </View>
    </Pressable>
  );
};

export { Button };
