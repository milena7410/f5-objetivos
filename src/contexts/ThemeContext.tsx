import * as React from "react";
import { useColorScheme, cssInterop } from "nativewind";
import AsyncStorage from "@react-native-async-storage/async-storage";

import "~/styles/global.css";

type Mode = "light" | "dark" | "system";
const ThemeContext = React.createContext<{
  isLoaded: boolean;
  colorScheme: "light" | "dark";
  toggleColorScheme: VoidFunction;
  setColorScheme(newColor: Mode): void;
  cssInterop: typeof cssInterop;
}>({
  isLoaded: false,
  colorScheme: "light",
  toggleColorScheme: () => {},
  setColorScheme: () => {},
  cssInterop: cssInterop,
});

export function ThemeProvider({ children }: React.PropsWithChildren) {
  const [colorMode, setColorMode] = React.useState<Mode>("system");
  const [isLoaded, setIsLoaded] = React.useState(false);
  const {
    colorScheme = "light",
    setColorScheme: setNativeColorSchema,
    toggleColorScheme: switchColorScheme,
  } = useColorScheme();

  React.useEffect(() => {
    AsyncStorage.getItem("theme").then((theme) => {
      setIsLoaded(true);
      if (theme) {
        const currentTheme = theme as typeof colorScheme;
        setColorScheme(currentTheme);
        setColorMode(currentTheme);
        return;
      }
      setColorScheme("system");
    });
  }, []);

  React.useEffect(() => {
    AsyncStorage.setItem("theme", colorMode);
  }, [colorMode]);

  const setColorScheme = React.useCallback((mode: Mode) => {
    if (mode === "system" && window?.matchMedia) {
      setColorMode(mode);
      mode = window.matchMedia("(prefers-color-scheme: dark)").matches
        ? "dark"
        : "light";
    }
    setNativeColorSchema(mode);
  }, []);

  const toggleColorScheme = React.useCallback(() => {
    switchColorScheme();
    const newColor = colorScheme === "light" ? "dark" : "light";
    setColorMode(newColor);
    AsyncStorage.setItem("theme", newColor);
  }, [colorScheme]);

  const value = React.useMemo(
    () => ({
      colorScheme,
      toggleColorScheme,
      setColorScheme,
      cssInterop,
      isLoaded,
    }),
    [colorScheme, toggleColorScheme, setColorScheme, isLoaded]
  );

  return (
    <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>
  );
}

export const useThemeContext = () => {
  return React.useContext(ThemeContext);
};
