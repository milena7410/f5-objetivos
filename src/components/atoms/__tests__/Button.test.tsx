import React from "react";
import {
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react-native";

import { useColorScheme } from "nativewind";

import { ThemeProvider } from "~/contexts/ThemeContext";
import { Button } from "../Button";
import "~/styles/global.css";

describe("Button Component", () => {
  const spyToggleColorScheme = jest.spyOn(
    useColorScheme(),
    "toggleColorScheme"
  );
  const spySetColorScheme = jest.spyOn(useColorScheme(), "setColorScheme");
  it("should render default button", async () => {
    render(
      <ThemeProvider>
        <Button />
      </ThemeProvider>
    );

    await waitFor(() => {
      const switchButton = screen.getByText("SWITCH");
      expect(spyToggleColorScheme).toHaveBeenCalledTimes(0);
      fireEvent.press(switchButton);
      expect(spyToggleColorScheme).toHaveBeenCalledTimes(1);

      const switchSystem = screen.getByText("SYSTEM");
      expect(spySetColorScheme).toHaveBeenCalledTimes(0);
      fireEvent.press(switchSystem);
      expect(spySetColorScheme).toHaveBeenCalledTimes(1);
      expect(screen.getByText("SWITCH")).toBeTruthy();
    });
  });
});
