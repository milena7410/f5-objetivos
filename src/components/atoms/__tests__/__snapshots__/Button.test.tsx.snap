// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component should button not be pressable when it is disabled 1`] = `
[
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className=" rounded-full self-center h-14 w-40 ""
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      className="  bg-gray-500 items-center justify-center rounded-full flex-1 self-stretch"
    >
      <Text
        className=" font-urbanist-bold text-white"
      >
        PRIMARY
      </Text>
    </View>
  </View>,
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className=" rounded-full self-center h-14 w-40 ""
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      className="  bg-gray-500 items-center justify-center rounded-full flex-1 self-stretch"
    >
      <Text
        className=" font-urbanist-bold text-white"
      >
        SECONDARY
      </Text>
    </View>
  </View>,
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className=" rounded-full self-center h-14 w-40 ""
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      className="  bg-gray-500 items-center justify-center rounded-full flex-1 self-stretch"
    >
      <Text
        className=" font-urbanist-bold text-white"
      >
        DANGER
      </Text>
    </View>
  </View>,
]
`;

exports[`Button Component should render button and pressable 1`] = `
[
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className=" rounded-full self-center h-14 w-40 ""
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      className="   bg-black dark:bg-white items-center justify-center rounded-full flex-1 self-stretch"
    >
      <Text
        className=" font-urbanist-bold text-white dark:text-black"
      >
        PRIMARY
      </Text>
    </View>
  </View>,
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className=" rounded-full self-center h-14 w-40 ""
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      className="   bg-black dark:bg-white items-center justify-center rounded-full flex-1 self-stretch"
    >
      <Text
        className=" font-urbanist-bold text-white dark:text-black"
      >
        SECONDARY
      </Text>
    </View>
  </View>,
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className=" rounded-full self-center h-14 w-40 ""
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      className="   bg-black dark:bg-white items-center justify-center rounded-full flex-1 self-stretch"
    >
      <Text
        className=" font-urbanist-bold text-white dark:text-black"
      >
        DANGER
      </Text>
    </View>
  </View>,
]
`;
